Source/BSGConfigurationBuilder.m

```diff
@@ -0,0 +1,145 @@
+#import "BSGConfigurationBuilder.h"
+#import "BugsnagConfiguration.h"
+
+NSString *const BSGAutoCollectBreadcrumbsKey = @"autoBreadcrumbs";
+NSString *const BSGAutoCaptureSessionsKey = @"autoSessions";
+
+/**
+ * Validate and set a value on config if the value is of the correct type.
+ * Remove the value from options after successful validation.
+ *
+ * @return config or nil if the value is not of the correct type
+ */
+BugsnagConfiguration *BSGValidateAndSetStringOption(BugsnagConfiguration *config, NSMutableDictionary *options, NSString *key);
+/**
+ * Validate and set a value on config if the value is a boolean.
+ * Remove the value from options after successful validation.
+ *
+ * @return config or nil if the value is a boolean
+ */
+BugsnagConfiguration *BSGValidateAndSetBooleanOption(BugsnagConfiguration *config, NSMutableDictionary *options, NSString *key);
+
+/**
+ * Validate and set notifyReleaseStages on config if the value is an array of
+ * strings.
+ * Remove the value from options after successful validation.
+ *
+ * @return config or nil if the value or contained values are not of the correct
+ * type
+ */
+BugsnagConfiguration *BSGValidateAndSetNotifyReleaseStages(BugsnagConfiguration *config, NSMutableDictionary *options);
+
+/**
+ * Validate and set notifyURL and sessionURL on config if the value is a
+ * dictionary containing exactly two values for keys "notify" and "sessions".
+ * Remove the value from options after successful validation.
+ *
+ * @return config or nil if the value or contained values are not of the correct
+ * type or name, to avoid data leakage to the hosted version from on-premise.
+ */
+BugsnagConfiguration *BSGValidateAndSetEndpoints(BugsnagConfiguration *config, NSMutableDictionary *options);
+
+@implementation BSGConfigurationBuilder
+
++ (BugsnagConfiguration *)configurationFromOptions:(NSDictionary *)options {
+    BugsnagConfiguration *config = [BugsnagConfiguration new];
+    NSMutableDictionary *properties = [options mutableCopy];
+    config = BSGValidateAndSetStringOption(config, properties, NSStringFromSelector(@selector(apiKey)));
```

Comment:
Why is `NSStringFromSelector(@selector(apiKey))` required rather than `@"apiKey"`?